File: homework2/warmup.js
exports.change = function(amount) {
    if (typeof amount !== 'number') {
        throw new TypeError('amount must be a number');
    } else if (amount < 0) {
        throw new RangeError('amount cannot be negative');
    }
    var QUARTER_VALUE = 25,
        DIME_VALUE = 10,
        NICKEL_VALUE = 5,
        quarters,
        dimes,
        nickels,
        pennies;

    quarters = Math.floor(amount / QUARTER_VALUE);
    amount %= QUARTER_VALUE;
    dimes = Math.floor(amount / DIME_VALUE);
    amount %= DIME_VALUE;
    nickels = Math.floor(amount / NICKEL_VALUE);
    pennies = amount % NICKEL_VALUE;
    return [quarters, dimes, nickels, pennies];
};

exports.stripQuotes = function(str) {
    return str.replace(/['"]/g, '');
};

exports.scramble = function(start) {
    // Fisher Yates shuffle from http://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle
    var result = start.split('');
    for (var i = 0; i < start.length; i++) {
        var j = Math.floor(Math.random() * i);
        var save = result[i];
        result[i] = result[j];
        result[j] = save;
    }
    return result.join('');
};

exports.powersOfTwo = function(limit, callback) {
    for (var x = 1; x <= limit; x *= 2) {
        callback(x);
    }
};

exports.prefixes = function(str, callback) {;
    for (var i = 0; i <= str.length; i++) {
        callback(str.substring(0, i));
    }
};

exports.interleave = function(a, b) {
    var aLen = a.length,
        bLen = b.length,
        maxLen = Math.max(aLen, bLen),
        result = [];
    for (var i = 0; i < maxLen; i++) {
        if (i < aLen) {
            result.push(a[i]);
        }
        if (i < bLen) {
            result.push(b[i]);
        }
    }
    return result;
};

exports.stutter = function(arr) {
    return this.interleave(arr, arr);
};

File: homework2/warmup.dart
library warmup;

import 'dart:collection';
import 'dart:math';

List<num> change(num amount) {
  if (amount < 0) {
    throw new RangeError('amount cannot be negative');
  }
  const QUARTER_VAL = 25, DIME_VAL = 10, NICKEL_VAL = 5;
  var quarters = amount ~/ QUARTER_VAL;
  amount = amount.remainder(QUARTER_VAL);
  var dimes = amount ~/ DIME_VAL;
  amount = amount.remainder(DIME_VAL);
  var nickels = amount ~/ NICKEL_VAL;
  var pennies = amount.remainder(NICKEL_VAL);
  return [quarters, dimes, nickels, pennies];
}

String stripQuotes(String begin) => begin.replaceAll(new RegExp("['\"]"), '');

String scramble(String begin) => (begin.split('')..shuffle()).join();

class _PowTwoIterator implements Iterator<int> {
  final int _itLimit;
  int currentPos;

  _PowTwoIterator(int limit): _itLimit = limit, currentPos = 0;

  bool moveNext() {
    currentPos *= 2;
    // Iterators in dart call moveNext before finding the first value, so we need to
    // start it off at the -1th position, and when moveNext is called, move it into the
    // first position. See comment above powersOfTwo.
    if (currentPos == 0) {
      currentPos = 1;
    }
    if (currentPos > _itLimit) {
      return false;
    }

    return true;
  }

  int get current => currentPos;
}

class _PowTwoIterable extends IterableBase<int> {
  int _itLimit;
  _PowTwoIterable(int limit) {
    _itLimit = limit;
  }

  Iterator<int> get iterator => new _PowTwoIterator(_itLimit);
}
// GODDAMMIT DART!!!!!!!
Iterable<int> powersOfTwo(int limit) {
  return new _PowTwoIterable(limit);
}

Iterable<String> prefixes(String toPrefix) =>
    new Iterable.generate(toPrefix.length, (index) => toPrefix.substring(0, index));

List interleave(List listA, List listB) {
  var maxLength = max(listA.length, listB.length);
  var returnList = [];
  for (var i = 0; i < maxLength; i++) {
    if (listA.length < maxLength) {
      returnList.add(listA[i]);
    }
    if (listB.length < maxLength) {
      returnList.add(listB[i]);
    }
  }
}

List stutter(List toStutter) => interleave(toStutter, toStutter);
